### ISSUE TITLE
Possibility to use dynamic_decode for training in OpenNMT 2.

### ISSUE THREAD

**Issue by atebbifakhr** [OpenNMT/OpenNMT-tf#537]
Hi,

Is it possible to update decoder weights based on loss computed over dynamic_decode output?

Thanks,

**Comment by guillaumekln** [comment]
Hi,

If I remember correctly, you had a similar question for OpenNMT-tf v1.  The situation is similar although the decoding code is now more contained and centralized.

Could you clarify how you would compute the loss? Right now I don't think `dynamic_decode` is differentiable.

**Comment by atebbifakhr** [comment]
Yes, actually I'm working on reinforcement learning and I need to back-propagate through predicted samples by model. 

In OpenNMT 1., `dynamic_decode` was differentiable if I wouldn't use `beam_search`. But in OpenNMT 2., `GreedySearch` and `BeamSearch` are using the same loop, which is not differentiable.

My loss is computed by multiplying the log probability of the generated translation with an externally computed reward. So I need to have a differentable log probability  computed by `dynamic_decode`.

**Comment by guillaumekln** [comment]
Do you know which part of the loop is not differentiable? Would removing `back_prop=False` on `tf.while_loop` be enough?

**Comment by atebbifakhr** [comment]
The value of variables in decoder are changing without changing anything in the code. I'm afraid there is something wrong in my code, because that flag `back_prop=False` should prevent any updates in the model.

**Comment by guillaumekln** [comment]
Do you have some updates on this?

I'm not even sure `back_prop=False` makes any difference in TensorFlow 2.0. Maybe we should just remove it.

**Comment by atebbifakhr** [comment]
I didn't see any difference. In both cases the back-propagation is being done.

**Comment by guillaumekln** [comment]
I removed this argument for good measure. I'm not sure what it does exactly.

So it seems we can close this issue if the back-propagation is working?

**Comment by atebbifakhr** [comment]
yes, Thanks a lot!
